### PR TITLE
Rollup until September 23

### LINKED ISSUE #315: Next merge with upstream
... should be done soon. I'll do it on the weekend if nobody else gets to it.

### PR DESCRIPTION
**Pull request by birkenfeld**
Another potpourri merge:

* Several trivial conflicting annoyances (URLs got the http -> https treatment).

* New members in the Lisp_Buffer struct which caused segfaults - we really should add tests to verify sizes of core types or switch to `bindgen`.

* Changes to how plist errors are reported.

* The `tempname` gnulib module is now used again, so I had to restore it in Remacs. Replacement by a Rust module can follow later.

* A big C change was made in the hashing functions, changed how to get the input string from different types. I kept the new big C function for now; a followup task will be to take the old functions in Rust code (that are now unused) and reuse/rewrite them to get an equivalent of `extract_data_from_obect`. (#323)

* A change to how emacs_globals is maintained in Rust; this was triggered by a new global variable being introduced and accessed in Rust code. TLDR: the old way was broken; now the globals struct is generated from the (also generated) header file.

Fixes #315.

### CODE REVIEW

**Comment by birkenfeld** [comment]
BTW this is now Emacs 27 :smile:

**Comment by birkenfeld** [comment]
Tests fail on OSX - can someone with a Mac please handle that?

**Comment by DavidDeSimone** [comment]
I'll be out for a couple hours, but if no one else has looked at OSX before I get back I can take a look.

File: rust_src/remacs-sys/lib.rs [hunk 1]
```diff
@@ -23,6 +23,7 @@ use libc::{c_char, c_uchar, c_short, c_int, c_double, c_float, c_void, ptrdiff_t
 
 
 include!(concat!(env!("OUT_DIR"), "/definitions.rs"));
+include!(concat!(env!("OUT_DIR"), "/globals.rs"));
```
Comment (Wilfred) [inline rust_src/remacs-sys/lib.rs]: The comment on line 6 should probably mention `emacs_globals` too, unless I'm missing something.

**Comment by Wilfred** [comment]
I'll have a closer look tomorrow, but everything I've read so far looks excellent. Thanks for tackling this.

Do you have any ideas for making merges easier? I'd be happy with some MS-DOS code slipping back in if it makes tracking upstream easier. Alternatively, perhaps we should focus porting the code that had `#ifdef MSDOS` in it?

If anyone else is interesting in reviewing the conflicts (rather than reviewing all the code written by upstream), I've been doing the following:

```diff
# do a three-way diff
$ git diff 853da4b27d67d7 853da4b27d67d7^1 853da4b27d67d7^2
...
diff --cc test/lisp/vc/vc-bzr-tests.el
index ac97d77167,e751f56286..8d83c4b50f
--- a/test/lisp/vc/vc-bzr-tests.el
+++ b/test/lisp/vc/vc-bzr-tests.el
@@@ -29,10 -29,7 +29,10 @@@
  (require 'vc-dir)
  
  (ert-deftest vc-bzr-test-bug9726 ()
-   "Test for http://debbugs.gnu.org/9726 ."
+   "Test for https://debbugs.gnu.org/9726 ."
 +  ;; Skipping on remacs until we figure out what's wrong.
 +  ;; https://github.com/Wilfred/remacs/issues/99
 +  (skip-unless (equal invocation-name "emacs"))
```

Note the different indentations for code from Remacs vs upstream vs fixing the conflict.

**Comment by DavidDeSimone** [comment]
So the failing OSX tests are also failing on mainline emacs for me. My results are:

```
Failures

REMACS:
  src/fileio-tests.log
  lisp/net/tramp-tests.log
  lisp/files-tests.log


MAINLINE:
  src/fileio-tests.log
  lisp/progmodes/flymake-tests.log
  lisp/net/tramp-tests.log
  lisp/files-tests.log
```

**Comment by birkenfeld** [comment]
@DavidDeSimone  Oh! So we can either fix the tests in Remacs and push the fix upstream (anyone got a CA?), disable them on Macs, or wait with the merge. (I'd hate the third option, as it basically voids a lot of my work here.)

**Comment by DavidDeSimone** [comment]
I like the idea of disable the tests of Macs and reporting this upstream.

**Comment by Wilfred** [comment]
Upstream's CI is showing Darwin as failed: https://hydra.nixos.org/jobset/gnu/emacs-trunk/evals

I'm happy with any solution proposed: fix on Remacs (I have CA to help push upstream) or disable for now. Agreed waiting is not really what we want.

File: rust_src/src/eval.rs [hunk 2]
```diff
@@ -66,21 +49,22 @@ macro_rules! error {
...
-    ($arg1:expr, $arg2:expr) => {{
-        xsignal!(::remacs_sys::Qargs_out_of_range, $arg1, $arg2);
-    }};
-    ($arg1:expr, $arg2:expr, $arg3:expr) => {{
-        xsignal!(::remacs_sys::Qargs_out_of_range, $arg1, $arg2, $arg3);
-    }}
+    ($($tt:tt)+) => { xsignal!(::remacs_sys::Qargs_out_of_range, $($tt)+); };
+}
+
+macro_rules! list {
```
Comment (Wilfred) [inline rust_src/src/eval.rs]: This is neat :)

**Comment by birkenfeld** [comment]
Thanks for the review! The test skips are now tracked in #328.

**Pull request merged**